File: examples/Ruby/lpserver.rb
#!/usr/bin/ruby

# Author: Han Holl <han.holl@pobox.com>

require 'rubygems'
require 'zmq'

class LPServer
  def initialize(connect)
    @ctx = ZMQ::Context.new(1)
    @socket = @ctx.socket(ZMQ::REP)
    @socket.bind(connect)
  end

  def run
    begin
      loop do
        rsl = yield @socket.recv
        @socket.send rsl
      end
    ensure
      @socket.close
      @ctx.close
    end
  end
      
end

if $0 == __FILE__
  cycles = 0
  srand
  LPServer.new(ARGV[0] || "tcp://*:5555").run do |request|
    cycles += 1
    if cycles > 3
      if rand(3) == 0
        puts "I: simulating a crash"
        break
      elsif rand(3) == 0
        puts "I: simulating CPU overload"
        sleep(3)
      end
    end
    puts "I: normal request (#{request})"
    sleep(1)
    request
  end
    
end

  

File: examples/Rust/lpserver.rs
use rand::{thread_rng, Rng};
use std::time::Duration;

fn main() {
    let context = zmq::Context::new();
    let server = context.socket(zmq::REP).unwrap();
    server.bind("tcp://*:5555").unwrap();

    let mut i = 0;
    loop {
        i += 1;
        let request = server.recv_msg(0).unwrap();
        println!("Got Request: {request:?}");
        server.send(request, 0).unwrap();
        std::thread::sleep(Duration::from_secs(1));

        if (i > 3) && (thread_rng().gen_range(0..3) == 0) {
            // simulate a crash
            println!("Oh no! Server crashed.");
            break;
        }
        if (i > 3) && (thread_rng().gen_range(0..3) == 0) {
            // simulate overload
            println!("Server is busy.");
            std::thread::sleep(Duration::from_secs(2));
        }
    }
}
